homeassistant/components/androidtv/media_player.py

```diff
@@ -168,180 +159,107 @@
-        config[CONF_DEVICE_CLASS],
-        10.0,
-        signer,
-    )
+    host = config[CONF_HOST]
 
-    if not aftv.available:
-        # Determine the name that will be used for the device in the log
-        if CONF_NAME in config:
-            device_name = config[CONF_NAME]
-        elif config[CONF_DEVICE_CLASS] == DEVICE_ANDROIDTV:
-            device_name = "Android TV device"
-        elif config[CONF_DEVICE_CLASS] == DEVICE_FIRETV:
-            device_name = "Fire TV device"
-        else:
-            device_name = "Android TV / Fire TV device"
-
-        _LOGGER.warning(
-            "Could not connect to %s at %s %s", device_name, address, adb_log
+    # get main data
+    config_data = {
+        CONF_HOST: host,
+        CONF_DEVICE_CLASS: config.get(CONF_DEVICE_CLASS, DEFAULT_DEVICE_CLASS),
+        CONF_PORT: config.get(CONF_PORT, DEFAULT_PORT),
+    }
+    for key in (CONF_ADBKEY, CONF_ADB_SERVER_IP, CONF_ADB_SERVER_PORT):
+        if key in config:
+            config_data[key] = config[key]
+
+    # get options
+    config_options = {}
+    for key in (
+        CONF_APPS,
+        CONF_EXCLUDE_UNNAMED_APPS,
+        CONF_GET_SOURCES,
+        CONF_SCREENCAP,
+        CONF_STATE_DETECTION_RULES,
+        CONF_TURN_OFF_COMMAND,
+        CONF_TURN_ON_COMMAND,
+    ):
+        if key in config:
+            if key == CONF_STATE_DETECTION_RULES:
+                config_options[key] = json.dumps(config[key])
+            else:
+                config_options[key] = config[key]
+
+    # save option to use with entry
+    if config_options:
+        hass.data.setdefault(DOMAIN, {}).setdefault(MIGRATION_DATA, {})[
+            host
+        ] = config_options
+
+    # Launch config entries setup
+    hass.async_create_task(
+        hass.config_entries.flow.async_init(
+            DOMAIN, context={"source": SOURCE_IMPORT}, data=config_data
         )
-        raise PlatformNotReady
+    )
 
-    async def _async_close(event):
-        """Close the ADB socket connection when HA stops."""
-        await aftv.adb_close()
 
-    # Close the ADB connection when HA stops
-    hass.bus.async_listen_once(EVENT_HOMEASSISTANT_STOP, _async_close)
+async def async_setup_entry(
+    hass: HomeAssistant,
+    entry: ConfigEntry,
+    async_add_entities: AddEntitiesCallback,
+) -> None:
+    """Set up the Android TV entity."""
+    aftv = hass.data[DOMAIN][entry.entry_id][ANDROID_DEV]
+    device_class = aftv.DEVICE_CLASS
+    device_name = "Android TV " if device_class == DEVICE_ANDROIDTV else "Fire TV "
+    device_name += entry.data[CONF_HOST]
 
     device_args = [
         aftv,
-        config[CONF_NAME],
-        config[CONF_APPS],
-        config[CONF_GET_SOURCES],
-        config.get(CONF_TURN_ON_COMMAND),
-        config.get(CONF_TURN_OFF_COMMAND),
-        config[CONF_EXCLUDE_UNNAMED_APPS],
-        config[CONF_SCREENCAP],
+        device_name,
+        entry.entry_id,
+        entry.unique_id,
     ]
 
```

Comment:
I implemented a different approach, trying to migrate existing entity with old unique ID, but may be yours solution is better. Please check and let me know.

Reply:
Your approach works the same as mine. I think the members will want your approach if it uses `async def async_migrate_entry(hass: HomeAssistant, config_entry: ConfigEntry) -> bool:` That way HA knows it supports migration.

Reply:
I think that I cannot use `async_migrate_entry` here because the `entity` is still not related to the `entry`. 
Anyway this seems working a could avoid breaking change, let's way for members review.

Reply:
My mistake. Take a look here: https://github.com/home-assistant/core/blob/dev/homeassistant/components/powerwall/__init__.py#L47

Reply:
My mistake. Take a look here: https://github.com/home-assistant/core/blob/dev/homeassistant/components/powerwall/__init__.py#L47

Reply:
Yes, I already see this code. But `async_migrate_entry` try to upgrade entity that are already related to the entry. Because the relation is created on first setup, it does not work.